File: 278-first-bad-version/278-first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int low = 1, ans = 0;
        int high = n;
        unsigned int mid = 0;
        
        while(low <= high){
            mid = low;
            mid += high;
            mid /= 2;
            
            if(isBadVersion(mid)){
                high = mid - 1;
                ans = mid;
            }
            else if(!isBadVersion(mid))
                low = mid + 1;
        }
        return ans;
    }
};

File: 278-first-bad-version/278-first-bad-version.py
# The isBadVersion API is already defined for you.
# @param version, an integer
# @return a bool
# def isBadVersion(version):

class Solution(object):
    def firstBadVersion(self, n):
        low, high = 1, n;
        
        while(low <= high):
            mid = high + (low - high)/2;
            
            if(isBadVersion(mid) == False):
                low = mid + 1;
            else:
                high = mid - 1;
        return low;
        
        